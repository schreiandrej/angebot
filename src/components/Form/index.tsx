import {
  initialFormState,
  mengezuschlagOptions,
  tankvolumenOptions,
} from '@/utils/variables'
import { useState } from 'react'
import { useForm } from 'react-hook-form'
import {
  InputADRZuschlag,
  InputDieselzuschlag,
  InputFüllstand,
  InputLiefermenge,
  InputPreis,
  InputTankvolumen,
  InputVorkasse,
  InputZuschlag,
} from './Inputs'
import { OutputSection } from './OutputSection'
import { IOptionsType, IForm, IFormData } from '../../types'
import { ButtonSubmit, ButtonDelete } from '../Buttons'
import { setStateOnSubmit } from '@/utils/setState'

export const FormComponent = () => {
  const [mengenzuschlag, setMengenzuschlag] = useState<IOptionsType>(
    mengezuschlagOptions[0]
  )
  const [tankvolumen, setTankvolumen] = useState<IOptionsType>(
    tankvolumenOptions[1]
  )
  const [formState, setFormState] = useState<IForm>(initialFormState)

  const {
    handleSubmit,
    reset,
    register,
    control,
    formState: { errors },
    clearErrors,
  } = useForm()

  const clearForm = () => {
    reset()
    clearErrors(['liter', 'preis'])

    setFormState(initialFormState)
  }

  const onSubmit = (data: any) => {
    if (data)
      setFormState(
        setStateOnSubmit(formState, data, tankvolumen, mengenzuschlag)
      )
  }

  return (
    <form className='flex flex-col gap-8' onSubmit={handleSubmit(onSubmit)}>
      <section className='flex flex-col gap-5'>
        <div className='flex flex-row w-full gap-3'>
          <div className='flex flex-col w-full gap-2'>
            <InputVorkasse register={register} errors={errors} />
            <InputPreis register={register} errors={errors} />
          </div>
          <div className='flex flex-col w-full gap-2'>
            <InputLiefermenge register={register} />
            <InputFüllstand register={register} />
          </div>
        </div>
        <div className='flex flex-col w-full gap-5'>
          <div className='flex flex-row gap-5'>
            <div className='flex flex-row w-full gap-3'>
              <InputZuschlag
                control={control}
                mengenzuschlag={mengenzuschlag}
                setMengenzuschlag={setMengenzuschlag}
              />
              <InputTankvolumen
                control={control}
                tankvolumen={tankvolumen}
                setTankvolumen={setTankvolumen}
              />
            </div>
            <div className='flex flex-row w-full justify-end gap-3'>
              <InputADRZuschlag register={register} />
              <InputDieselzuschlag register={register} />
            </div>
          </div>
        </div>
      </section>
      <OutputSection formState={formState} />
      <section className='flex flex-row gap-3 '>
        <ButtonSubmit />
        <ButtonDelete deleteResults={clearForm} />
      </section>
    </form>
  )
}
