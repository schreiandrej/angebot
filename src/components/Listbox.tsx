import { Listbox, Transition } from '@headlessui/react'

type OptionsType = {
  id: number
  name: string
  value: number
  unavailable: boolean
}

type ListboxComponentProps = {
  options: OptionsType[]
  selectedOption: OptionsType
  setSelectedOption: any
  className?: string
}

export const ListboxComponent = ({
  options,
  selectedOption,
  setSelectedOption,
}: ListboxComponentProps) => {
  return (
    <Listbox
      as='div'
      value={selectedOption.value}
      onChange={(e) => {
        setSelectedOption(e)
      }}
      className='relative w-full'
    >
      {({ open }) => (
        <>
          <Listbox.Button
            className={`w-full border border-gray-400 hover:border-hover focus:outline-none active:ring-gray-600 focus:ring-ring-gray-600 focus:ring-1 active:ring-1 rounded-lg p-1`}
          >
            {selectedOption.name}
          </Listbox.Button>
          {/* Use the Transition + open render prop argument to add transitions. */}
          <Transition
            show={open}
            enter='transition duration-100 ease-out'
            enterFrom='transform scale-95 opacity-0'
            enterTo='transform scale-100 opacity-100'
            leave='transition duration-75 ease-out'
            leaveFrom='transform scale-300 opacity-100'
            leaveTo='transform scale-95 opacity-0'
          >
            <Listbox.Options
              static
              className='absolute w-full text-center focus:outline-none bg-white overflow-visible'
            >
              {options.map((option) => (
                <Listbox.Option
                  key={option.name}
                  value={option}
                  disabled={option.unavailable}
                  className='cursor-pointer hover:text-gray-400 my-3'
                >
                  {option.name !== selectedOption.name && option.name}
                </Listbox.Option>
              ))}
            </Listbox.Options>
          </Transition>
        </>
      )}
    </Listbox>
  )
}

export const CheckmarkIcon = ({ className }: { className?: string }) => {
  return (
    <svg
      xmlns='http://www.w3.org/2000/svg'
      fill='none'
      viewBox='0 0 24 24'
      stroke='currentColor'
      width='10'
      height='10'
      className={`${className}`}
    >
      <path
        strokeLinecap='round'
        strokeLinejoin='round'
        strokeWidth={2}
        d='M5 13l4 4L19 7'
      />
    </svg>
  )
}
