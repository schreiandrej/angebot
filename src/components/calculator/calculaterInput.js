import { Input } from '@/components/base/forms/input'

export const CalcInput = ({ className, register, clearForm }) => {
  return (
    <div className='relative w-full'>
      <Input
        register={register}
        type='text'
        name='calc'
        autoComplete='off'
        inputStyles='w-full py-4 text-center'
        autoFocus
        placeholder='1 + 1'
      />
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        stroke='currentColor'
        width='20'
        height='20'
        className='absolute right-3 absoluteY-center z-10 cursor-pointer text-gray-700 hover:text-gray-500'
        onClick={clearForm}
      >
        <path
          strokeLinecap='round'
          strokeLinejoin='round'
          strokeWidth={2}
          d='M10 14l2-2m0 0l2-2m-2 2l-2-2m2 2l2 2m7-2a9 9 0 11-18 0 9 9 0 0118 0z'
        />
      </svg>
    </div>
  )
}
