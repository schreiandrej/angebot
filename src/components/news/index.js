import { Container } from '@/components/base/container'
import { Carousel } from '@/components/base/carousel'

export const News = ({ articles, className }) => {
  return (
    <Container className={`${className}`}>
      <Carousel>
        {articles.map((article, index) => {
          return (
            <h3 className='min-w-full px-4 text-center' key={index}>
              <a
                className='hover:text-hover'
                href={`https://www.presseportal.de${article.url}`}
              >
                {article.text}
              </a>
            </h3>
          )
        })}
      </Carousel>
    </Container>
  )
}
