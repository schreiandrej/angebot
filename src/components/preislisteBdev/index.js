import { Title } from '@/components/base/title'
import { useState, useEffect } from 'react'
import { Container } from '../base/container'
import { ListboxComponent } from '../base/headlessUI/listbox'

// { id: 1, name: 'Durward Reynolds', value: x, unavailable: false }

export const SelectPreislisteBdev = ({ preisebdev }) => {
  const [selectedOption, setSelectedOption] = useState(preisebdev[0])

  if (!selectedOption) return <div>loading....</div>

  return (
    <Container className='flex flex-col w-full items-center gap-2'>
      <ListboxComponent
        options={preisebdev}
        selectedOption={selectedOption}
        setSelectedOption={setSelectedOption}
      />
      <div>{`${selectedOption.plz}`}</div>
      <div>{`${selectedOption.value}`}</div>
    </Container>
  )
}

export const PreislisteBdev = ({ preisebdev }) => {
  const convert = (numberString) =>
    new Intl.NumberFormat('de-DE', {
      style: 'currency',
      currency: 'EUR',
    }).format(parseFloat(numberString.replace(',', '.')))
  return (
    <div className='text-lg w-full'>
      <Title className='text-2xl mb-6'>Preise des BDEV: </Title>
      <ul>
        {preisebdev.map((item, index) => (
          <li
            key={index}
            className='flex flex-row items-center justify-between'
          >
            <span className='flex-grow text-heading'>
              {Object.keys(item)[0]}
            </span>
            <span className='flex-shrink text-xs'>
              {item[Object.keys(item)[0]].plz.split(',').length > 4
                ? item[Object.keys(item)[0]].plz.slice(-22)
                : item[Object.keys(item)[0]].plz}
            </span>
            <span className='flex-shrink  pl-8'>
              {convert(item[Object.keys(item)[0]].netto1000)}
            </span>
          </li>
        ))}
      </ul>
    </div>
  )
}
