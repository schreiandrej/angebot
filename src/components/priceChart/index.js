import { useState, useEffect } from 'react'
import Link from 'next/link'
import { Line } from 'react-chartjs-2'
import { useForm } from 'react-hook-form'
import DatePicker from 'react-datepicker'
import { sub, format } from 'date-fns'
import { options } from '@/components/priceChart/options'
import { getData } from '@/components/priceChart/sortData'
import { getPostleitzahlArray } from '@/components/priceChart/getPostleitzeitArray'
import { Title } from '@/components/base/title'
import { Container } from '@/components/base/container'
import { Select } from '@/components/base/forms/select'

export const LineChart = ({ className, preisliste }) => {
  const { register, handleSubmit, errors } = useForm()
  const [chartData, setChartData] = useState({
    labels: '',
    values: '',
    date: '',
    preis: '',
  })
  const [searchedDate, setSearchedDate] = useState(new Date())

  const data = {
    labels: chartData.labels,
    datasets: [
      {
        data: chartData.values,
        fill: false,
        backgroundColor: 'rgb(255, 99, 132)',
        borderColor: 'rgba(217, 119, 32, 0.2)',
      },
    ],
  }

  const onSubmit = (data) => {
    const lineData = getData(preisliste, data.postleitzahlSelect, searchedDate)
    setChartData({
      values: lineData.values,
      labels: lineData.labels,
      date: lineData.date,
      plz: lineData.postleitzahl,
      preis: lineData.preis,
      today: lineData.today,
    })
  }

  const optionsArray = [{ option: '', value: '' }]
  getPostleitzahlArray(preisliste).map((item) =>
    optionsArray.push({ option: item, value: item })
  )

  useEffect(() => {
    const lineData = getData(preisliste, 30, searchedDate)
    setChartData({
      values: lineData.values,
      labels: lineData.labels,
      date: lineData.date,
      plz: lineData.postleitzahl,
      preis: lineData.preis,
      today: lineData.today,
    })
  }, [])

  return (
    <Container className={` ${className} relative`}>
      <Title>preisentwicklung</Title>
      <Line data={data} options={options} />
      <form
        onSubmit={handleSubmit(onSubmit)}
        className=' flex flex-col w-full items-center gap-1 text-xs pt-10 '
      >
        <div className='relative flex w-full gap-1'>
          <div className='relative'>
            <label className='absolute -top-5 left-4'>Datum</label>
            <DatePicker
              selected={searchedDate}
              onChange={(date) => setSearchedDate(date)}
              closeOnScroll={true}
              dateFormat='dd.MM.yyyy'
              startDate={searchedDate}
              className='w-min-full'
            />
          </div>
          <Select
            className='h-full '
            name='postleitzahlSelect'
            options={optionsArray}
            label='Postleitzahl'
            register={register({
              required: 'Bitte eine Postleitzahl eingeben!',
            })}
            inputStyles='h-full px-2'
          />
          {errors?.postleitzahlSelect && (
            <p className='text-xs text-red-600 absolute -top-5 right-0 '>
              {errors.postleitzahlSelect.message}
            </p>
          )}
          <button
            type='submit'
            className='w-full button-outlined h-full py-2 important-my-0'
          >
            suchen
          </button>
        </div>
      </form>
      <div className='absolute flex flex-col text-md top-3 left-5 hover:text-gray-500'>
        {chartData.date}
        <Link href='/update-isn'>
          <a>ISN: {chartData.preis} Ct./l</a>
        </Link>
      </div>
    </Container>
  )
}
