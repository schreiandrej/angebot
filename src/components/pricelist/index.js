import { Title } from '@/components/base/title'
import { Container } from '@/components/base/container'
import { ListItem } from '@/components/pricelist/listItem'

export const Preise = ({ data, className }) => {
  const mitte1 = new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR',
  }).format(parseFloat(data.mitte1.replace(',', '.')))
  const mitte2 = new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR',
  }).format(parseFloat(data.mitte2.replace(',', '.')))
  const nordWest2 = new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR',
  }).format(parseFloat(data.nordWest2.replace(',', '.')))

  return (
    <Container className={`${className}`}>
      <Title>aktuelle preise bdev</Title>
      <ol className='w-full sm:px-20'>
        <ListItem gebiet='Mitte I' plz='30,31' price={mitte1} />
        <ListItem gebiet='Nord-West II' plz='32,33' price={nordWest2} />
        <ListItem gebiet='Mitte II' plz='38,39' price={mitte2} />
      </ol>
    </Container>
  )
}
