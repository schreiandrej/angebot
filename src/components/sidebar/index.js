import { useState } from 'react'
import { ISNPreis } from './isnPreis'
import format from 'date-fns/format'
import de from 'date-fns/locale/de'
import { PreislisteBdev } from './preislisteBdev'
import { useSortData } from '@/hooks/useSortData'
import { Container } from '@/components/base/container'
import { Title } from '@/components/base/title'

export const Sidebar = ({ preisebdev, preisliste, setShowContent }) => {
  const heute = new Date()

  const { postleitzahl, date, preis } = useSortData(preisliste, 33, heute)

  return (
    <Container className='items-start justify-between'>
      <Title className=''>
        <div className='text-3xl'>Guten Morgen!</div>
        <div className='text-2xl'>{format(heute, 'PPPP', { locale: de })}</div>
      </Title>
      <ISNPreis isnpreis={preis} setShowContent={setShowContent} />
      <PreislisteBdev preisebdev={preisebdev} />
    </Container>
  )
}
