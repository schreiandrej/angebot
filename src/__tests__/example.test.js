import { render, screen, fireEvent, cleanup, act } from '@testing-library/react'
import { Calculator } from '@/components/calculator'

afterEach(cleanup)

test('renders Calculator component', () => {
  const { asFragment } = render(<Calculator />)
  expect(asFragment()).toMatchSnapshot()
})

test('calculate 3+3', async () => {
  const { getByTestId } = render(<Calculator />)

  await act(async () => {
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: '3,254+3,25' },
    })
    fireEvent.click(await screen.getByRole('submit-button'))
  })
  expect(getByTestId('result')).toHaveTextContent('6,504')
})
