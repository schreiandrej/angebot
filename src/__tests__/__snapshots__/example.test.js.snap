// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Calculator component 1`] = `
<DocumentFragment>
  <div
    class="undefined flex flex-col w-full lg:h-full p-8 overflow-hidden border border-base rounded-md items-center justify-center relative "
  >
    <h2
      class="justify-self-start text-xl pb-3 font-semibold dark:font-normal text-heading uppercase"
    >
      calculator
    </h2>
    <form
      class="flex flex-col w-full h-full justify-between"
    >
      <div
        class="flex flex-col w-full"
      >
        <div
          class="relative"
        >
          <div
            class="relative w-full"
          >
            <div
              class="relative z-10"
            >
              <input
                autocomplete="off"
                class="z-10 text-sm cursor-text  w-full py-4 text-center"
                name="calc"
                placeholder="1 + 1"
                type="text"
              />
              <label
                class="absolute -z-10 left-4 flex items-center focus:outline-none transition-all duration-200 ease-in-out text-xs -top-5 undefined"
                for="calc"
              />
            </div>
            <svg
              class="absolute right-3 absoluteY-center z-10 cursor-pointer text-gray-700 hover:text-gray-500"
              fill="none"
              height="20"
              stroke="currentColor"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 14l2-2m0 0l2-2m-2 2l-2-2m2 2l2 2m7-2a9 9 0 11-18 0 9 9 0 0118 0z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
        <div
          class="flex flex-col w-full py-4 pl-3 justify-end items-end"
        />
      </div>
      <div
        class="flex flex-row gap-2 w-full lg:justify-end"
      >
        <button
          class="button-outlined group w-full undefined "
          role="submit-button"
          type="submit"
        >
          <svg
            class="text-center text-green-700 group-hover:text-green-300 w-full"
            fill="none"
            height="20"
            stroke="currentColor"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 7h6m0 10v-3m-3 3h.01M9 17h.01M9 14h.01M12 14h.01M15 11h.01M12 11h.01M9 11h.01M7 21h10a2 2 0 002-2V5a2 2 0 00-2-2H7a2 2 0 00-2 2v14a2 2 0 002 2z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
        <button
          class="button-outlined group undefined"
          type="button"
        >
          <svg
            class="text-center text-red-500 text-opacity-30 group-hover:text-opacity-70"
            fill="none"
            height="20"
            stroke="currentColor"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
